### PR TITLE
Add `WithPgWeb`

### LINKED ISSUE #3081: Consider adding support for pgweb as an alternative to pgAdmin
pgweb (https://github.com/sosedoff/pgweb) is another popular (>8K stars on GitHub) web-based admin dashboard for PostgreSQL that should consider adding support to `Aspire.Hosting.PostgreSQL`. It is [recommended by the Azure Container Apps PostgreSQL add-on docs](https://learn.microsoft.com/azure/container-apps/tutorial-dev-services-postgresql?tabs=bash#binding-to-the-dev-service).

image

### PR DESCRIPTION
**Pull request by Alirexaa**
In this PR, we add pgweb (https://github.com/sosedoff/pgweb) a web-based admin dashboard for PostgreSQL.

using `WithPgWeb`, We collect all registered `PostgresDatabaseResource` instances and create a configuration file per instance, then bind the location of these files to the pgweb container `bookmark` directory. for more information about pgweb bookmark you can see https://github.com/sosedoff/pgweb/wiki/Server-Connection-Bookmarks.

We used the Eventing API and subscribed `AfterEndpointsAllocatedEvent` to create configuration files.

We used pgweb `connect` API for functional testing to ensure pgweb is configured correctly.

```csharp
var testConnectionApiUrl = $"{pgWebEndpoint.Scheme}://{pgWebEndpoint.Host}:{pgWebEndpoint.Port}/api/connect";

var pipeline = new ResiliencePipelineBuilder().AddRetry(new Polly.Retry.RetryStrategyOptions
{
    Delay = TimeSpan.FromSeconds(2),
    MaxRetryAttempts = 10,
}).Build();

await pipeline.ExecuteAsync(async (ct) =>
{
    var httpContent = new MultipartFormDataContent
    {
        { new StringContent(dbName), "bookmark_id" }
    };

    client.DefaultRequestHeaders.Add("x-session-id", Guid.NewGuid().ToString());

    var response = await client.PostAsync(testConnectionApiUrl, httpContent, ct)
        .ConfigureAwait(false);

    response.EnsureSuccessStatusCode();
}, cts.Token).ConfigureAwait(false);
```

close: #3081 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5098)

### CODE REVIEW

File: Aspire.sln [hunk 1]
```diff
@@ -526,7 +527,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Hosting.PostgreSQL.T
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Hosting.Qdrant.Tests", "tests\Aspire.Hosting.Qdrant.Tests\Aspire.Hosting.Qdrant.Tests.csproj", "{8E2AA85E-C351-47B4-AF91-58557FAD5840}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Aspire.Hosting.Milvus.Tests", "tests\Aspire.Hosting.Milvus.Tests\Aspire.Hosting.Milvus.Tests.csproj", "{986886B7-0E38-4890-92C3-5B46DE322DAF}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Hosting.Milvus.Tests", "tests\Aspire.Hosting.Milvus.Tests\Aspire.Hosting.Milvus.Tests.csproj", "{986886B7-0E38-4890-92C3-5B46DE322DAF}"
```
Comment (mitchdenny) [inline Aspire.sln:530]: Why this change?

File: playground/PostgresEndToEnd/PostgresEndToEnd.AppHost/Program.cs [hunk 2]
```diff
@@ -21,6 +21,13 @@
 // External resources.
 var db10 = builder.AddPostgres("pg10").WithPgAdmin().PublishAsConnectionString().AddDatabase("db10");
 
+builder.Configuration["Parameters:pass"] = "p!ssw0rd1";
```
Comment (mitchdenny) [inline playground/PostgresEndToEnd/PostgresEndToEnd.AppHost/Program.cs:24]: Please use `p@ssw0rd1` as we already have a suppression setup for it.

File: src/Aspire.Hosting.PostgreSQL/PgAdminConfigWriterHook.cs [hunk 3]
```diff
@@ -58,3 +58,45 @@ public Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, C
         return Task.CompletedTask;
     }
 }
+
+internal sealed class PgWebConfigWriterHook : IDistributedApplicationLifecycleHook
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PgAdminConfigWriterHook.cs:62]: This can go into a separate file.

File: src/Aspire.Hosting.PostgreSQL/PgAdminConfigWriterHook.cs [hunk 4]
```diff
@@ -58,3 +58,45 @@ public Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, C
...
+                database = "{postgresDatabase.DatabaseName}"
+                sslmode = "require"
+                """;
+
+            if (!Directory.Exists(serverFileMount.Source!))
+            {
+                Directory.CreateDirectory(serverFileMount.Source!);
+            } 
+            var filePath = Path.Combine(serverFileMount.Source!, $"{postgresDatabase.Name}.toml");
+            File.WriteAllText(filePath, fileContent);
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PgAdminConfigWriterHook.cs:97]: async?

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 5]
```diff
@@ -114,6 +114,73 @@ public static IResourceBuilder<PgAdminContainerResource> WithHostPort(this IReso
...
+    public static IResourceBuilder<PgWebContainerResource> WithHostPort(this IResourceBuilder<PgWebContainerResource> builder, int? port)
+    {
+        return builder.WithEndpoint("http", endpoint =>
+        {
+            endpoint.Port = port;
+        });
+    }
+
+    /// <summary>
+    /// Adds an administration and development platform for PostgreSQL to the application model using pgweb. This version the package defaults to the 2.3-latest tag of the latest container image
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:132]: Can you make this more concise and split out any extended explanation into a `<remarks />` tag.

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 6]
```diff
@@ -114,6 +114,73 @@ public static IResourceBuilder<PgAdminContainerResource> WithHostPort(this IReso
...
+    /// </code>
+    /// </example>
+    /// <param name="builder">The Postgres server resource builder.</param>
+    /// <param name="configureContainer">Configuration callback for pgweb container resource.</param>
+    /// <param name="containerName">The name of the container (Optional).</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> WithPgWeb<T>(this IResourceBuilder<T> builder, Action<IResourceBuilder<PgWebContainerResource>>? configureContainer = null, string? containerName = null) where T : PostgresDatabaseResource
+    {
+
+        if (builder.ApplicationBuilder.Resources.OfType<PgWebContainerResource>().SingleOrDefault() is { } existingPgAdminResource)
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:155]: should be `existingPgWebResource`

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 7]
```diff
@@ -114,6 +114,73 @@ public static IResourceBuilder<PgAdminContainerResource> WithHostPort(this IReso
...
+                                               .WithImageRegistry(PostgresContainerImageTags.PgwebRegistry)
+                                               .WithHttpEndpoint(targetPort: 8081, name: "http")
+                                               .WithBindMount(dir, "/.pgweb/bookmarks")
+                                               .WithArgs("--bookmarks-dir=/.pgweb/bookmarks")
+                                               .WithArgs("--sessions")
+                                               .ExcludeFromManifest();
+
+            configureContainer?.Invoke(pgwebContainerBuilder);
+            return builder;
+
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:180]: remove extra line

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 8]
```diff
@@ -114,6 +114,73 @@ public static IResourceBuilder<PgAdminContainerResource> WithHostPort(this IReso
...
+    /// var postgres = builder.AddPostgres("postgres");
+    /// var db = postgres.AddDatabase("db")
+    ///   .WithPgWeb();  
+    /// var api = builder.AddProject&lt;Projects.Api&gt;("api")
+    ///   .WithReference(db);
+    ///  
+    /// builder.Build().Run(); 
+    /// </code>
+    /// </example>
+    /// <param name="builder">The Postgres server resource builder.</param>
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:148]: `<param />` elements should go under `<summary />` element.

File: src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs [hunk 9]
```diff
@@ -11,4 +11,7 @@ internal static class PostgresContainerImageTags
     public const string PgAdminRegistry = "docker.io";
     public const string PgAdminImage = "dpage/pgadmin4";
     public const string PgAdminTag = "8.8";
+    public const string PgwebRegistry = "docker.io";
```
Comment (mitchdenny) [inline src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs:14]: `Pgweb` or `PgWeb`?

**Comment by mitchdenny** [comment]
@Alirexaa this is looking pretty good. A few minor things. We'll need a docs issue/PR filed for this as well (probably just updating the existing doc page for Postgres).

**Comment by Alirexaa** [comment]
@mitchdenny besides these things, this PR needs more to do, right now bookmarks populate in pgweb but when selecting one of those to connect, connection refuse occur. 
Also need to write some tests

File: playground/PostgresEndToEnd/PostgresEndToEnd.AppHost/Program.cs (hunk 2 above)
Comment (Alirexaa) [reply playground/PostgresEndToEnd/PostgresEndToEnd.AppHost/Program.cs:24]: Its not url encoded.

File: Aspire.sln (hunk 1 above)
Comment (Alirexaa) [reply Aspire.sln:530]: I don't no, visual studio stuff :))

File: src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs (hunk 9 above)
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs:14]: `pgweb` is the correct one, but we could use `PgWeb`.

File: Aspire.sln (hunk 1 above)
Comment (sebastienros) [reply Aspire.sln:529]: I believe these are project types. These are both C# but the 9A ones uses the SDK project system (whatever that means). I assume it depends which version created it, and the newer ones try to update them.

**Comment by Alirexaa** [comment]
To pass all tests we need to import `docker.io/sosedoff/pgweb/latest` to `netaspireci.azurecr.io`

**Comment by Alirexaa** [comment]
@eerhardt could you please import `docker.io/sosedoff/pgweb/latest` to make ci green?

File: src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs [hunk 10]
```diff
@@ -11,4 +11,7 @@ internal static class PostgresContainerImageTags
     public const string PgAdminRegistry = "docker.io";
     public const string PgAdminImage = "dpage/pgadmin4";
     public const string PgAdminTag = "8.8";
+    public const string PgWebRegistry = "docker.io";
+    public const string PgWebImage = "sosedoff/pgweb";
+    public const string PgWebTag = "latest";
```
Comment (eerhardt) [inline src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs:16]: We've been not using `latest` tags as much as possible due to non-determinism. See examples:

* https://github.com/dotnet/aspire/issues/4828
* https://github.com/dotnet/aspire/issues/5057

File: tests/Aspire.Hosting.PostgreSQL.Tests/AddPostgresTests.cs [hunk 11]
```diff
@@ -376,6 +376,48 @@ public void WithPgAdminAddsContainer()
...
+    {
+        var builder = DistributedApplication.CreateBuilder();
+        builder.AddPostgres("mypostgres1").WithPgWeb();
+        builder.AddPostgres("mypostgres2").WithPgWeb();
+
+        Assert.Single(builder.Resources.OfType<PgWebContainerResource>());
+    }
+
+    [Fact]
+    public void WithRedisInsightSupportsChangingContainerImageValues()
```
Comment (eerhardt) [inline tests/Aspire.Hosting.PostgreSQL.Tests/AddPostgresTests.cs:390]:
```suggestion
    public void WithPgWebSupportsChangingContainerImageValues()
```

File: src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Aspire.Hosting.Postgres;
+
+internal sealed class PgWebConfigWriterHook : IDistributedApplicationLifecycleHook
+{
+    public async Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken)
+    {
+        var adminResource = appModel.Resources.OfType<PgWebContainerResource>().Single();
+        var serverFileMount = adminResource.Annotations.OfType<ContainerMountAnnotation>().Single(v => v.Target == "/.pgweb/bookmarks");
+        var postgresInstances = appModel.Resources.OfType<PostgresDatabaseResource>();
```
[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs:16]: do we need to introduce a new annotation to make a relation between `PgWebContainerResource` and `PostgresDatabaseResource ` or maybe something like this

```csharp

public sealed class PgWebContainerResource(string name,PostgresDatabaseResource postgresDatabaseResource) : ContainerResource(name)
{
   public PostgresDatabaseResource PostgresDatabaseResource { get; } = postgresDatabaseResource;
}
```

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 13]
```diff
@@ -114,6 +114,75 @@ public static IResourceBuilder<PgAdminContainerResource> WithHostPort(this IReso
...
+            var builderForExistingResource = builder.ApplicationBuilder.CreateResourceBuilder(existingPgWebResource);
+            configureContainer?.Invoke(builderForExistingResource);
+            return builder;
+        }
+        else
+        {
+            builder.ApplicationBuilder.Services.TryAddLifecycleHook<PgWebConfigWriterHook>();
+
+            containerName ??= $"{builder.Resource.Name}-pgweb";
+            var dir = Directory.CreateTempSubdirectory().FullName;
```
[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: do we have a hook or event when a resource stopped/finished?

I think we should add one API for that in `IDistributedApplicationLifecycleHook`

File: src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs (hunk 10 above)
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs]: now we use the `0.15.0` tag,Could you please import  `docker.io/sosedoff/pgweb:0.15.0`?

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: This is actually a problem right now. @karolz-ms interested in your thoughts here. We sometimes drop temporary files in disk and I'd like to be able to try and ensure that they are cleaned up. We can probably have some kind of temporary file register that goes and does clean up but its highly likely that some of those files/directories are going to be locked because things like containers are spinning down.

This might be one of those things that DCP could help with?

[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: Yes, I think DCP would be a good place for logic that deals with "cleanup" tasks like removing temporary log files. 

There are probably many ways to do this, but the simplest one that will work today, is to put them inside the DCP session folder (which the app host runtime knows about). That folder is automatically deleted when DCP is shutting down. Just make sure the name is reasonably unique when creating the file.

[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: (we might be missing an API for the extensions that will tell them what the "temporary folder for the current application run" is)

[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: I think it would be good to be able to provide "cleanup paths" along with each resource when it shuts down. The path could either be a specific file or a whole directory.

[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: We could do that too. Feel free to open an issue and we can add it to our backlog for DCP.

[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:170]: what is our approach for cleanup temp files for this PR?

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:226]: @mitchdenny - would this be a good usage of the new [Eventing API](https://github.com/dotnet/aspire/pull/5223)? We could have an event that fires when the resource is shut down.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:226]: > what is our approach for cleanup temp files for this PR?

Since we already have the same issue with pgadmin:

https://github.com/dotnet/aspire/blob/4b4ebce8ef3f8db3fd43f0838d6734b46609f469/src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs#L101-L111

I'd be fine with continuing here to just create temp files and leave them, for now, until we get the clean up capability.

File: playground/PostgresEndToEnd/PostgresEndToEnd.AppHost/Program.cs [hunk 14]
```diff
@@ -21,6 +21,9 @@
 // External resources.
 var db10 = builder.AddPostgres("pg10").WithPgAdmin().PublishAsConnectionString().AddDatabase("db10");
 
+var db11 = builder.AddPostgres("pg11").WithPgWeb().AddDatabase("postgres");
+var db12 = builder.AddPostgres("pg12").WithPgWeb().AddDatabase("db12");
```
Comment (eerhardt) [inline playground/PostgresEndToEnd/PostgresEndToEnd.AppHost/Program.cs:25]: Do we really need 2 of them?

File: src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs [hunk 15]
```diff
@@ -0,0 +1,39 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Lifecycle;
+
+namespace Aspire.Hosting.Postgres;
+
+internal sealed class PgWebConfigWriterHook : IDistributedApplicationLifecycleHook
```
[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs:9]: I will.

File: src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs:15]: @eerhardt, what about this?
We have same behavior for pgadmin.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs:15]: I guess if the behavior is the same as pgadmin, then I'm fine. If we feel we should have a different behavior for both - log an issue.

**Comment by radical** [comment]
Can you please add some information in PR description, maybe even just links to pgweb?

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs:226]: Yes ... although it wouldn't be 100%. If someone kills the process via the debugger (most common scenario) I don't think we guarantee that this is run. But better than nothing I guess.

File: src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.PostgreSQL/PgWebConfigWriterHook.cs:15]: We have an open issue for that. see #2060.

**Comment by eerhardt** [comment]
> @eerhardt could you please import `docker.io/sosedoff/pgweb/latest` to make ci green?

I've pushed it to the netaspireci.azurecr.io registry. However, I don't believe the test will pass because the CI machines don't have Docker Desktop installed, which means there isn't network communication between the containers.

**Comment by Alirexaa** [comment]
> > @eerhardt could you please import `docker.io/sosedoff/pgweb/latest` to make ci green?
> 
> I've pushed it to the netaspireci.azurecr.io registry. However, I don't believe the test will pass because the CI machines don't have Docker Desktop installed, which means there isn't network communication between the containers.

What can I do for that?

File: src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs (hunk 10 above)
Comment (radical) [reply src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs]: I can do that.

**Comment by eerhardt** [comment]
> What can I do for that?

For now we've been disabling the test in CI until the problem can be resolved.

File: src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs (hunk 10 above)
Comment (radical) [reply src/Aspire.Hosting.PostgreSQL/PostgresContainerImageTags.cs]: @eerhardt already did it!

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for this contribution!
